Create task for evaluating rating systems

jauggy opened the pull request:
This task when run will give the brier score to assess how good the win prediction function is. To run use this:
```
mix teiserver.prediction_stats
```

This will give the brier score as a log in the console. Lower score is better.
![image](https://github.com/user-attachments/assets/a09c403b-d4b7-4a59-b540-63a7a8e68a12)

So if we predict that team 1 will win with 60% with probability, if they do win the brier score is
```
(0.6-1)^2
```

We can repeat this for many matches.

The win prediction function for openskill is:
```
phi_major((mu_a - mu_b) / :math.sqrt(n * betasq + sigma_sq_a + sigma_sq_b))
```

phi_major is the cumulative distribution function of a standard normal. phi_major(0) = 50%. In other words, the OS library assumes 50% win predict when team skill is equal on both teams (this is the numerator).

When we balance games, we balance on team rating. BAR doesn't have a win predict function but we can approximate it using:
```
phi_major((team_rating_a - team_rating_b) / :math.sqrt(n * betasq + sigma_sq_a + sigma_sq_b))
```
In other words, if the rating of teams are equal BAR assumes 50% win probability.

If we use this approximation for win prediction for BAR's rating system, what would be its brier score? Is it better than Openskill's library brier score?

The results of this task will return the brier score using Openskill's win predict as well as our BAR-adjusted win predict.

jauggy commented:
@L-e-x-o-n when you have spare time can you please run this task on integration server and tell me the log from console. Thanks.

L-e-x-o-n commented:
```ex
brier_score: %{
  noob_matches: %{
    num_matches: 736,
    bar_brier: 0.2821430560962619,
    openskill_brier: 0.288875576717513,
    provisional_brier: 0.3280238734364291
  },
  pro_matches: %{
    num_matches: 264,
    bar_brier: 0.27385031089673,
    openskill_brier: 0.2751992706687806,
    provisional_brier: 0.2751992706687806
  }
}
```

jauggy commented:
Wow that was not expected. I was expecting the provisional rating to outperform for noob matches.

jauggy commented:
@L-e-x-o-n when you have time can you run again and post the results from the logs. 

I've added two new rating systems:
```
      :provisional_20 ->
        min(num_matches / 20, 1) * (player.skill - player.uncertainty)

      :provisional_10 ->
        min(num_matches / 10, 1) * (player.skill - player.uncertainty)
```

And want to see the brier score for these.

L-e-x-o-n commented:
Looks like you are not outputting the results anywhere.

jauggy commented:
> Looks like you are not outputting the results anywhere.

Oops. Added logs back. Please try again.

jauggy commented:
@L-e-x-o-n 
I've added another scoring function called logloss. This one penalises incorrect confidence more than Brier score.

Can you please run this task
```
mix teiserver.logloss_stats
```
and tell me the results? Thanks.

jauggy commented:
Actually ignore that. I'm reading the article and I think my interpretation is wrong.

jauggy commented:
Actually just rereading some other articles and I think my interpretation is correct. So can you please run this:
```
mix teiserver.logloss_stats
```
and tell me the results. Thanks.

jauggy commented:
@L-e-x-o-n I've modified it to only use team size 8 and also added in provisional 5 rating.

When you have time please run this (one at a time) and provide the logs
```
mix teiserver.brier_stats
```
```
mix teiserver.logloss_stats
```

I also assume season 2 matches are not in integration. If they are I will need to modify the script to filter by season.

jauggy commented:
## Results
The results indicate that starting new players at 0 does not improve win prediction since the scores are higher. I may need to do more analysis of individual matches in the future to understand what is going on.

```
brier_score: %{
  noob_matches: [
    %{
      num_matches: 1414,
      rating_system: :openskill,
      brier_score: 0.2601494331433799
    },
    %{num_matches: 1414, rating_system: :bar, brier_score: 0.2520541588481419},
    %{
      num_matches: 1414,
      rating_system: :provisional_5,
      brier_score: 0.2681889344201138
    },
    %{
      num_matches: 1414,
      rating_system: :provisional_10,
      brier_score: 0.2735037093920037
    },
    %{
      num_matches: 1414,
      rating_system: :provisional_20,
      brier_score: 0.27452238035972093
    }
  ],
  experienced_matches: [
    %{
      num_matches: 586,
      rating_system: :openskill,
      brier_score: 0.25554770765000867
    },
    %{num_matches: 586, rating_system: :bar, brier_score: 0.2533347051900348},
    %{
      num_matches: 586,
      rating_system: :provisional_5,
      brier_score: 0.2534170091974334
    },
    %{
      num_matches: 586,
      rating_system: :provisional_10,
      brier_score: 0.2534170091974334
    },
    %{
      num_matches: 586,
      rating_system: :provisional_20,
      brier_score: 0.2534170091974334
    }
  ]
}
```
```
log_loss_score: %{
  noob_matches: [
    %{
      num_matches: 1414,
      rating_system: :openskill,
      log_loss_score: 0.728104860956009
    },
    %{
      num_matches: 1414,
      rating_system: :bar,
      log_loss_score: 0.7043006154456513
    },
    %{
      num_matches: 1414,
      rating_system: :provisional_5,
      log_loss_score: 0.7511890421815198
    },
    %{
      num_matches: 1414,
      rating_system: :provisional_10,
      log_loss_score: 0.7664910511687886
    },
    %{
      num_matches: 1414,
      rating_system: :provisional_20,
      log_loss_score: 0.7780222119255443
    }
  ],
  experienced_matches: [
    %{
      num_matches: 586,
      rating_system: :openskill,
      log_loss_score: 0.7100900510888444
    },
    %{num_matches: 586, rating_system: :bar, log_loss_score: 0.706734491694382},
    %{
      num_matches: 586,
      rating_system: :provisional_5,
      log_loss_score: 0.7069780841701404
    },
    %{
      num_matches: 586,
      rating_system: :provisional_10,
      log_loss_score: 0.7069780841701404
    },
    %{
      num_matches: 586,
      rating_system: :provisional_20,
      log_loss_score: 0.7069780841701404
    }
  ]
}
```

jauggy commented:
Starting new players at 0 makes prediction worse, but I have a theory to why. The balance algorithm tends to put the weakest noob with the best captain. We might be undervaluing the captain's contribution to the game, i.e. in 8v8 game the captain contributes more than 1/8th of the team's total.

To test this theory I have now doubled the contribution of the captain when assessing win prediction.

@L-e-x-o-n 

When you have time please run this (one at a time) and provide the logs
```
mix teiserver.brier_stats
```
```
mix teiserver.logloss_stats
```

jauggy commented:
# Results
```
log_loss_score: %{
  noob_matches: [
    %{
      num_matches: 1414,
      rating_system: :openskill,
      log_loss_score: 0.78961678867669
    },
    %{
      num_matches: 1414,
      rating_system: :bar,
      log_loss_score: 0.7646677309495655
    },
    %{
      num_matches: 1414,
      rating_system: :provisional_5,
      log_loss_score: 0.8054379606018283
    },
    %{
      num_matches: 1414,
      rating_system: :provisional_10,
      log_loss_score: 0.8179802019208307
    },
    %{
      num_matches: 1414,
      rating_system: :provisional_20,
      log_loss_score: 0.8256590637011004
    }
  ],
  experienced_matches: [
    %{
      num_matches: 586,
      rating_system: :openskill,
      log_loss_score: 0.7605550564824592
    },
    %{num_matches: 586, rating_system: :bar, log_loss_score: 0.7541571294907663},
    %{
      num_matches: 586,
      rating_system: :provisional_5,
      log_loss_score: 0.7542522972488848
    },
    %{
      num_matches: 586,
      rating_system: :provisional_10,
      log_loss_score: 0.7542522972488848
    },
    %{
      num_matches: 586,
      rating_system: :provisional_20,
      log_loss_score: 0.7542522972488848
    }
  ]
}
```
```
brier_score: %{
  noob_matches: [
    %{
      num_matches: 1414,
      rating_system: :openskill,
      brier_score: 0.2780715647167735
    },
    %{num_matches: 1414, rating_system: :bar, brier_score: 0.27146739655906793},
    %{
      num_matches: 1414,
      rating_system: :provisional_5,
      brier_score: 0.28517445544276604
    },
    %{
      num_matches: 1414,
      rating_system: :provisional_10,
      brier_score: 0.2895568327823006
    },
    %{
      num_matches: 1414,
      rating_system: :provisional_20,
      brier_score: 0.2896346684377928
    }
  ],
  experienced_matches: [
    %{
      num_matches: 586,
      rating_system: :openskill,
      brier_score: 0.27403219063827106
    },
    %{num_matches: 586, rating_system: :bar, brier_score: 0.2713302119503804},
    %{
      num_matches: 586,
      rating_system: :provisional_5,
      brier_score: 0.27134458641164344
    },
    %{
      num_matches: 586,
      rating_system: :provisional_10,
      brier_score: 0.27134458641164344
    },
    %{
      num_matches: 586,
      rating_system: :provisional_20,
      brier_score: 0.27134458641164344
    }
  ]
}
```

# Conclusion
The results are even worse. Now I have no idea what is the reason. I think the next step is to create a script to get the ids of badly predicted matches.

jauggy commented:
@L-e-x-o-n I completely messed up the code in the past as it was assuming team 0 was always the winner. When you have time can you please run these commands one at a time and send the results:

```
mix teiserver.prediction_stats
```
```
mix teiserver.prediction_stats --doublecaptain
```

L-e-x-o-n commented:
```ex
score: %{
  noob_matches: [
    %{score: 0.3149699280229058, num_matches: 1414, rating_system: :openskill},
    %{score: 0.2899979994291328, num_matches: 1414, rating_system: :bar},
    %{
      score: 0.2728631321210531,
      num_matches: 1414,
      rating_system: :provisional_5
    },
    %{
      score: 0.2664406306714633,
      num_matches: 1414,
      rating_system: :provisional_10
    },
    %{
      score: 0.26005415948499166,
      num_matches: 1414,
      rating_system: :provisional_20
    }
  ],
  experienced_matches: [
    %{score: 0.2669931715038141, num_matches: 586, rating_system: :openskill},
    %{score: 0.25968243682552666, num_matches: 586, rating_system: :bar},
    %{
      score: 0.26012955449961744,
      num_matches: 586,
      rating_system: :provisional_5
    },
    %{
      score: 0.26012955449961744,
      num_matches: 586,
      rating_system: :provisional_10
    },
    %{
      score: 0.26012955449961744,
      num_matches: 586,
      rating_system: :provisional_20
    }
  ]
}
```

```ex
opts: [doublecaptain: true]
score: %{
  noob_matches: [
    %{score: 0.32509418088589387, num_matches: 1414, rating_system: :openskill},
    %{score: 0.30089850962755726, num_matches: 1414, rating_system: :bar},
    %{
      score: 0.28279971283966127,
      num_matches: 1414,
      rating_system: :provisional_5
    },
    %{
      score: 0.27583558655342294,
      num_matches: 1414,
      rating_system: :provisional_10
    },
    %{
      score: 0.2684935157467026,
      num_matches: 1414,
      rating_system: :provisional_20
    }
  ],
  experienced_matches: [
    %{score: 0.2771501489924787, num_matches: 586, rating_system: :openskill},
    %{score: 0.269634059257517, num_matches: 586, rating_system: :bar},
    %{
      score: 0.2700027693543584,
      num_matches: 586,
      rating_system: :provisional_5
    },
    %{
      score: 0.2700027693543584,
      num_matches: 586,
      rating_system: :provisional_10
    },
    %{
      score: 0.2700027693543584,
      num_matches: 586,
      rating_system: :provisional_20
    }
  ]
}
```

jauggy commented:
Thanks. Now the results are making a bit more sense.

For noob matches (any player with less than 20 matches), the rating systems from best to worse are:
Provisional 20 (brier: 0.260)
Provisional 10
Provisional 5 
BAR (brier: 0.29)
openskill (brier: 0.315)

For experienced matches  (no noobs), the rating systems from best to worse are:
BAR i.e. max(0, skill-uncertainty)
skill-uncertainty
skill

Still the brier score being higher than 0.25 is not good. It means that just predicting 50% for every match would have given a better brier score.

jauggy commented:
@L-e-x-o-n  when you have time can you please run again the below and post results:
```
mix teiserver.prediction_stats
```

This now includes `provisional_30` and `provisional_50`. While it does seem like a high number to calibrate a player, I did see Dota 2 requires 100 hours to unlock ranked so it's not too ridiculous.

jauggy commented:
@L-e-x-o-n I've added ways to filter out matches without parties.

We you have time please run:
```
mix teiserver.prediction_stats --offset=10 --noparties
```
This filters out matches where any team has a party.

```
mix teiserver.prediction_stats --offset=10 --equalparties
```
This one filters out matches where number of players in parties on each team is not equal

jauggy commented:
@L-e-x-o-n I've added ability to filter by team size. Maybe we can get better accuracy with smaller games? When you have time please run
```
mix teiserver.prediction_stats --offset=10 --minteamsize=2 --maxteamsize=5
```

L-e-x-o-n commented:
> @L-e-x-o-n I've added ability to filter by team size. Maybe we can get better accuracy with smaller games? When you have time please run
> 
> ```
> mix teiserver.prediction_stats --offset=10 --minteamsize=2 --maxteamsize=5
> ```

```
** (MatchError) no match of right hand side value: [0]
    (teiserver 0.1.0) lib/teiserver/mix_tasks/prediction_stats.ex:209: Mix.Tasks.Teiserver.PredictionStats.get_match_error/2
    (elixir 1.18.2) lib/enum.ex:1714: Enum."-map/2-lists^map/1-1-"/2
    (elixir 1.18.2) lib/enum.ex:1714: Enum."-map/2-lists^map/1-1-"/2
    (teiserver 0.1.0) lib/teiserver/mix_tasks/prediction_stats.ex:70: Mix.Tasks.Teiserver.PredictionStats.run/1
    iex:1: (file)
    ```

jauggy commented:
I think I messed up. This query is pulling PvE games too probably:
```
    select distinct  tbm.id, tbm.inserted_at  from
    teiserver_battle_matches tbm
    inner join teiserver_game_rating_logs tgrl
    on tgrl.match_id = tbm.id
    and tbm.team_size >= $1
    and tbm.team_size <= $2
    and tbm.team_count = $3
    and tgrl.value is not null
    and tgrl.season = 1
    order by tbm.inserted_at DESC
    limit $4;
```

I'll need to make changes.

jauggy commented:
Results without offset:
```
opts: [minteamsize: 2, maxteamsize: 5]
score: %{
  noob_matches: [
    %{
      score: 0.2651119464506306,
      num_matches: 1955,
      matches_with_win_data: 1955,
      rating_system: :openskill
    },
    %{
      score: 0.24904255537532272,
      num_matches: 1955,
      matches_with_win_data: 1955,
      rating_system: :bar
    },
    %{
      score: 0.24919787950604064,
      num_matches: 1955,
      matches_with_win_data: 1955,
      rating_system: :bar_negative
    },
    %{
      score: 0.23986428012185693,
      num_matches: 1955,
      matches_with_win_data: 1955,
      rating_system: :leaderboard_2
    },
    %{
      score: 0.23742227090581453,
      num_matches: 1955,
      matches_with_win_data: 1955,
      rating_system: :leaderboard_3
    },
    %{
      score: 0.24789154734603266,
      num_matches: 1955,
      matches_with_win_data: 1955,
      rating_system: :provisional_5
    },
    %{
      score: 0.24718051176900305,
      num_matches: 1955,
      matches_with_win_data: 1955,
      rating_system: :provisional_10
    },
    %{
      score: 0.24983247142785206,
      num_matches: 1955,
      matches_with_win_data: 1955,
      rating_system: :provisional_20
    },
    %{
      score: 0.25482564381590994,
      num_matches: 1955,
      matches_with_win_data: 1955,
      rating_system: :provisional_30
    },
    %{
      score: 0.261636465405348,
      num_matches: 1955,
      matches_with_win_data: 1955,
      rating_system: :provisional_50
    }
  ],
  experienced_matches: [
    %{
      score: 0.19199503349281677,
      num_matches: 45,
      matches_with_win_data: 45,
      rating_system: :openskill
    },
    %{
      score: 0.18874277331005643,
      num_matches: 45,
      matches_with_win_data: 45,
      rating_system: :bar
    },
    %{
      score: 0.18875277680518868,
      num_matches: 45,
      matches_with_win_data: 45,
      rating_system: :bar_negative
    },
    %{
      score: 0.1865308679711122,
      num_matches: 45,
      matches_with_win_data: 45,
      rating_system: :leaderboard_2
    },
    %{
      score: 0.18523763431180135,
      num_matches: 45,
      matches_with_win_data: 45,
      rating_system: :leaderboard_3
    },
    %{
      score: 0.18875277680518868,
      num_matches: 45,
      matches_with_win_data: 45,
      rating_system: :provisional_5
    },
    %{
      score: 0.18875277680518868,
      num_matches: 45,
      matches_with_win_data: 45,
      rating_system: :provisional_10
    },
    %{
      score: 0.18875277680518868,
      num_matches: 45,
      matches_with_win_data: 45,
      rating_system: :provisional_20
    },
    %{
      score: 0.18875277680518868,
      num_matches: 45,
      matches_with_win_data: 45,
      rating_system: :provisional_30
    },
    %{
      score: 0.18875277680518868,
      num_matches: 45,
      matches_with_win_data: 45,
      rating_system: :provisional_50
    }
  ]
}
```

# Summary
I didn't post results with offset but they were worse. Leaderboard rating is still the best rating method. Out of the provisional rating systems, provisional_10 is the best.

Leaderboard rating still has the issue that players start at 50th percentile.
